src/commands/classified/external.rs

```diff
@@ -467,76 +469,94 @@ fn spawn(
 
     trace!(target: "nu::run::external", "built command {:?}", process);
 
+    // TODO Switch to async_std::process once it's stabilized
     if let Ok(mut child) = process.spawn() {
-        let stream = async_stream! {
-            if let Some(mut input) = input {
-                let mut stdin_write = child.stdin
+        let (tx, rx) = mpsc::sync_channel(0);
+
+        let mut stdin = child.stdin.take();
+
+        let stdin_write_tx = tx.clone();
+        let stdout_read_tx = tx;
+        let stdin_name_tag = command.name_tag.clone();
+        let stdout_name_tag = command.name_tag;
+
+        std::thread::spawn(move || {
```

Comment:
We spawn two threads: one to write to stdin, the other to read from stdout. It's very important that these happen concurrently, otherwise an infinite stream will result in us writing to stdin  indefinitely.

When they happen concurrently, something that can eliminate an infinite loop (e.g., `first`, `head`) will complete, dropping our sender/receiver channels, which will cascade back through the pipeline, shutting everything down and ensuring all of these threads terminate.